### ISSUE TITLE
Only booleans are allowed in a negated boolean false positiive

### ISSUE THREAD

**Issue by jakubvojacek** [efabrica-team/phpstan-latte#34]
AFAIK this shouldnt be reported

```
{varType array<Motv\ApiModule\Entity\ReportScheduleAttachementEntity> $reports}

<ul style="list-style-type: none">
{foreach $reports as $report}
    <li>
        <a class="synchronous" n:href="download!, $report->getId()"><i class="fa fa-file-pdf-o"></i> {basename($report->getPath())}</a>
    </li>
{/foreach}
</ul>

{if count($reports) === 0}
    <p>There are no reports to be downloaded.</p>
{/if}
```

reports 

```
 ------ ------------------------------------------------------------------------------------ 
  Line   FrontModule/Components/Config/ReportScheduleHistoryControl.latte                    
 ------ ------------------------------------------------------------------------------------ 
  11     Only booleans are allowed in a negated boolean, Latte\Runtime\Template|null given.  
 ------ ------------------------------------------------------------------------------------ 
```

generated code:

```
<?php

declare (strict_types=1);
use Latte\Runtime as LR;
final class PHPStanLatteTemplate extends Latte\Runtime\Template
{
    public function main(array $ʟ_args) : void
    {
        extract($this->params);
        /** @var string $baseUrl */
        /** @var string $basePath */
        /** @var Latte\Runtime\FilterInfo $ʟ_fi */
        /** @var Nette\Security\User $user */
        /** @var Nette\Application\UI\Presenter $presenter */
        /** @var Nette\Application\UI\Control $control */
        /** @var array<stdClass> $flashes */
        /** @var Nette\Localization\Translator $netteLocalizationTranslatorFilter */
        /** @var array<Motv\ApiModule\Entity\ReportScheduleAttachementEntity> $reports */
        /** @var Motv\FrontModule\Components\Config\ReportScheduleHistoryControl $actualClass */
        /** @var Motv\FrontModule\Components\Config\ReportScheduleHistoryControl $control */
        unset($ʟ_args);
        if ($this->global->snippetDriver?->renderSnippets($this->blocks[self::LayerSnippet], $this->params)) {
            return;
        }
        echo '
<ul style="list-style-type: none">
';
        /* line 4 */
        foreach ($reports as $report) {
            echo '    <li>
        <a href="';
            $actualClass->handleDownload($report->getId());
            echo '" class="synchronous"><i class="fa fa-file-pdf-o"></i> ';
            /* line 6 */
            echo LR\Filters::escapeHtmlText(basename($report->getPath()));
            echo '</a>
    </li>
';
        }
        echo '</ul>

';
        /* line 11 */
        if (count($reports) === 0) {
            echo '    <p>There are no reports to be downloaded.</p>
';
        }
    }
    public function prepare() : array
    {
        extract($this->params);
        /** @var string $baseUrl */
        /** @var string $basePath */
        /** @var Latte\Runtime\FilterInfo $ʟ_fi */
        /** @var Nette\Security\User $user */
        /** @var Nette\Application\UI\Presenter $presenter */
        /** @var Nette\Application\UI\Control $control */
        /** @var array<stdClass> $flashes */
        /** @var Nette\Localization\Translator $netteLocalizationTranslatorFilter */
        /** @var array<Motv\ApiModule\Entity\ReportScheduleAttachementEntity> $reports */
        /** @var Motv\FrontModule\Components\Config\ReportScheduleHistoryControl $actualClass */
        /** @var Motv\FrontModule\Components\Config\ReportScheduleHistoryControl $control */
        if (!$this->getReferringTemplate() || $this->getReferenceType() === 'extends') {
            foreach (array_intersect_key(['report' => '4'], $this->params) as $ʟ_v => $ʟ_l) {
                trigger_error("Variable \${$ʟ_v} overwritten in foreach on line {$ʟ_l}");
            }
        }
        return get_defined_vars();
    }
}
```

**Comment by lulco** [comment]
Oh! I see, it's probably this line:
```
if (!$this->getReferringTemplate() || $this->getReferenceType() === 'extends') {
```
It is something latte generates here, we will add it to ignored errors.

Thanks for report

**Comment by MartinMystikJonas** [comment]
@lulco FYI these are ignore error patterns we use on compiled latte templates to avoid false positives in code generated by latte

```
    - '#Parameter \#1 \$array of function extract is passed by reference, so it expects variables only\.#'
    - '#Closure invoked with 1 parameter, 0 required\.#'
    - '#Offset .* on array.* in isset\(\) does not exist\.#'
    - "#Access to private property \\$blocks of parent class Latte\\\\Runtime\\\\Template\\.#"
    - "#Method Nette\\\\Application\\\\UI\\\\.?Renderable::redrawControl\\(\\) invoked with 2 parameters, 0 required\\.#"
    - "|Parameter \\#1 \\$var_array of function extract is passed by reference, so it expects variables only\\.|"
```

**Comment by lulco** [comment]
@MartinMystikJonas thanks, can you prepare PR? Here is the right place https://github.com/efabrica-team/phpstan-latte/blob/main/src/Error/ErrorBuilder.php#L18

also, do you have an idea why we need so many backslashes there? :D

**Comment by MartinMystikJonas** [comment]
Well we do not need so much with '' but it has to be doible escaped in "". Which is fact I learned after I had most ignor pattetns already written.

**Comment by jakubvojacek** [comment]
Should I ignore this error when reported by phpstan or do you want me to report it here @lulco ? It happens quite often on weird lines, eg 

- `{snippet}`
- `<tr><td><div class="cell-cover">{$case->getValue()}</div></td><td><div class="cell-cover">{$case->getLabel()}</div></td></tr>`
- `{else}`

**Comment by lulco** [comment]
We can try to change this condition
```
if (!$this->getReferringTemplate() || $this->getReferenceType() === 'extends') {
```
to
```
if ($this->getReferringTemplate() === null || $this->getReferenceType() === 'extends') {
```

**Comment by lulco** [comment]
What is the php code for your snippet?

**Comment by jakubvojacek** [comment]
you mean in /tmp folder?

**Comment by lulco** [comment]
yes

**Comment by jakubvojacek** [comment]
eg 

```
{varType array<int, Motv\ApiModule\Entity\RecognitionTemplateImageEntity> $templateImages}
{varType array<Motv\ApiModule\Entity\RecognitionTrainingImageEntity> $images}
{varType Motv\ApiModule\Entity\RecognitionModelEntity $model}
{snippet}
	<h2>{$model->getName()} <span class="badge badge-default">{$model->getStatus()->getLabel()}</span></h2>
	{if count($images) === 0}
		<p>There are no training images collected yet.</p>
	{/if}
	<div style="text-align: center;" n:snippet="images">
		{var $processedImages = []}
        {foreach $images as $image}
	        {if !in_array($image->getTemplateImageId(), $processedImages, true)}
                {var $processedImages[] = $image->getTemplateImageId()}
		        <h1>Section {$templateImages[$image->getTemplateImageId()]->getName()} ({$counts[$image->getTemplateImageId()]})</h1>
		        <p>You should be looking for following image: <img style="max-width: 80px;" src="{$templateImages[$image->getTemplateImageId()]->getImage()}"/></p>
	        {/if}
			<img src="{$image->getImage()}" style="max-width: 45%; float: right;"/>
	        <img src="{$image->getImageOriginal()}" style="max-width:45%; float: left"/>
	        <div style="clear: both;"></div>
	        {if in_array($model->getStatus(), [Motv\ApiModule\Enum\RecognitionModelStatusEnum::CREATED, Motv\ApiModule\Enum\RecognitionModelStatusEnum::TRAINING_SET_COLLECTION, Motv\ApiModule\Enum\RecognitionModelStatusEnum::TRAINING_SET_COMPLETED], true)}
	            <span style="float: right;"><a n:href="removeImage, id => $image->getId()" class="btn-danger btn ajax" data-naja-history="off"><span
			        class="motv-icon motv-icon-trash"></span> Delete</a></span>
	        {/if}
            <ul>
	            <li>
		            {$image->getDescription()} - ({$image->getX1()}, {$image->getY1()}), ({$image->getX2()}, {$image->getY2()})
	            </li>
            </ul>

            {sep}
		        <hr/>
		        <hr/>
		        <hr/>{/sep}
        {/foreach}
	</div>
{/snippet}
```

```php
<?php

use Latte\Runtime as LR;
final class PHPStanLatteTemplate extends Latte\Runtime\Template
{
    public const Blocks = ['snippet' => ['' => 'block1', 'images' => 'blockImages']];
    public function main(array $ʟ_args) : void
    {
        extract($this->params);
        /** @var string $baseUrl */
        /** @var string $basePath */
        /** @var Latte\Runtime\FilterInfo $ʟ_fi */
        /** @var Nette\Security\User $user */
        /** @var Nette\Application\UI\Presenter $presenter */
        /** @var Nette\Application\UI\Control $control */
        /** @var array<stdClass> $flashes */
        /** @var Nette\Localization\Translator $netteLocalizationTranslatorFilter */
        /** @var mixed $templateImages */
        /** @var mixed $counts */
        /** @var mixed $images */
        /** @var mixed $model */
        /** @var Motv\FrontModule\Components\Tools\TrainingImagesControl $actualClass */
        /** @var Motv\FrontModule\Components\Tools\TrainingImagesControl $control */
        unset($ʟ_args);
        if ($this->global->snippetDriver?->renderSnippets($this->blocks[self::LayerSnippet], $this->params)) {
            return;
        }
        echo '<div id="', htmlspecialchars($this->global->snippetDriver->getHtmlId('')), '">';
        /* line 4 */
        $this->renderBlock('', [], null, 'snippet');
        echo '</div>';
    }
    public function prepare() : array
    {
        extract($this->params);
        /** @var string $baseUrl */
        /** @var string $basePath */
        /** @var Latte\Runtime\FilterInfo $ʟ_fi */
        /** @var Nette\Security\User $user */
        /** @var Nette\Application\UI\Presenter $presenter */
        /** @var Nette\Application\UI\Control $control */
        /** @var array<stdClass> $flashes */
        /** @var Nette\Localization\Translator $netteLocalizationTranslatorFilter */
        /** @var mixed $templateImages */
        /** @var mixed $counts */
        /** @var mixed $images */
        /** @var mixed $model */
        /** @var Motv\FrontModule\Components\Tools\TrainingImagesControl $actualClass */
        /** @var Motv\FrontModule\Components\Tools\TrainingImagesControl $control */
        if (!$this->getReferringTemplate() || $this->getReferenceType() === 'extends') {
            foreach (array_intersect_key(['image' => '11'], $this->params) as $ʟ_v => $ʟ_l) {
                trigger_error("Variable \${$ʟ_v} overwritten in foreach on line {$ʟ_l}");
            }
        }
        return get_defined_vars();
    }
    /** {snippet} on line 4 */
    public function block1(array $ʟ_args) : void
    {
        extract($this->params);
        /** @var string $baseUrl */
        /** @var string $basePath */
        /** @var Latte\Runtime\FilterInfo $ʟ_fi */
        /** @var Nette\Security\User $user */
        /** @var Nette\Application\UI\Presenter $presenter */
        /** @var Nette\Application\UI\Control $control */
        /** @var array<stdClass> $flashes */
        /** @var Nette\Localization\Translator $netteLocalizationTranslatorFilter */
        /** @var mixed $templateImages */
        /** @var mixed $counts */
        /** @var mixed $images */
        /** @var mixed $model */
        /** @var Motv\FrontModule\Components\Tools\TrainingImagesControl $actualClass */
        /** @var Motv\FrontModule\Components\Tools\TrainingImagesControl $control */
        unset($ʟ_args);
        /* line 4 */
        $this->global->snippetDriver->enter('', 'static');
        try {
            echo '
	<h2>';
            /* line 5 */
            echo LR\Filters::escapeHtmlText($model->getName());
            echo ' <span class="badge badge-default">';
            /* line 5 */
            echo LR\Filters::escapeHtmlText($model->getStatus()->getLabel());
            echo '</span></h2>
';
            /* line 6 */
            if (count($images) === 0) {
                echo '		<p>There are no training images collected yet.</p>
';
            }
            echo '	<div style="text-align: center;"';
            echo ' id="', htmlspecialchars($this->global->snippetDriver->getHtmlId('images')), '"';
            echo '>';
            /* line 9 */
            $this->renderBlock('images', [], null, 'snippet');
            echo '</div>
';
        } finally {
            $this->global->snippetDriver->leave();
        }
    }
    /** n:snippet="images" on line 9 */
    public function blockImages(array $ʟ_args) : void
    {
        extract($this->params);
        /** @var string $baseUrl */
        /** @var string $basePath */
        /** @var Latte\Runtime\FilterInfo $ʟ_fi */
        /** @var Nette\Security\User $user */
        /** @var Nette\Application\UI\Presenter $presenter */
        /** @var Nette\Application\UI\Control $control */
        /** @var array<stdClass> $flashes */
        /** @var Nette\Localization\Translator $netteLocalizationTranslatorFilter */
        /** @var mixed $templateImages */
        /** @var mixed $counts */
        /** @var mixed $images */
        /** @var mixed $model */
        /** @var Motv\FrontModule\Components\Tools\TrainingImagesControl $actualClass */
        /** @var Motv\FrontModule\Components\Tools\TrainingImagesControl $control */
        unset($ʟ_args);
        /* line 9 */
        $this->global->snippetDriver->enter('images', 'static');
        try {
            echo "\n";
            /* line 10 */
            $processedImages = [];
            /* line 11 */
            foreach ($iterator = $ʟ_it = new Latte\Essential\CachingIterator($images, $ʟ_it ?? null) as $image) {
                /* line 12 */
                if (!in_array($image->getTemplateImageId(), $processedImages, true)) {
                    /* line 13 */
                    $processedImages[] = $image->getTemplateImageId();
                    echo '		        <h1>Section ';
                    /* line 14 */
                    echo LR\Filters::escapeHtmlText($templateImages[$image->getTemplateImageId()]->getName());
                    echo ' (';
                    /* line 14 */
                    echo LR\Filters::escapeHtmlText($counts[$image->getTemplateImageId()]);
                    echo ')</h1>
		        <p>You should be looking for following image: <img style="max-width: 80px;" src="';
                    /* line 15 */
                    echo LR\Filters::escapeHtmlAttr(LR\Filters::safeUrl($templateImages[$image->getTemplateImageId()]->getImage()));
                    echo '"/></p>
';
                }
                echo '			<img src="';
                /* line 17 */
                echo LR\Filters::escapeHtmlAttr(LR\Filters::safeUrl($image->getImage()));
                echo '" style="max-width: 45%; float: right;"/>
	        <img src="';
                /* line 18 */
                echo LR\Filters::escapeHtmlAttr(LR\Filters::safeUrl($image->getImageOriginal()));
                echo '" style="max-width:45%; float: left"/>
	        <div style="clear: both;"></div>
';
                /* line 20 */
                if (in_array($model->getStatus(), [Motv\ApiModule\Enum\RecognitionModelStatusEnum::CREATED, Motv\ApiModule\Enum\RecognitionModelStatusEnum::TRAINING_SET_COLLECTION, Motv\ApiModule\Enum\RecognitionModelStatusEnum::TRAINING_SET_COMPLETED], true)) {
                    echo '	            <span style="float: right;"><a href="';
                    echo LR\Filters::escapeHtmlAttr($this->global->uiControl->link('removeImage', ['id' => $image->getId()]));
                    echo '" class="btn-danger btn ajax" data-naja-history="off"><span
			        class="motv-icon motv-icon-trash"></span> Delete</a></span>
';
                }
                echo '            <ul>
	            <li>
		            ';
                /* line 26 */
                echo LR\Filters::escapeHtmlText($image->getDescription());
                echo ' - (';
                /* line 26 */
                echo LR\Filters::escapeHtmlText($image->getX1());
                echo ', ';
                /* line 26 */
                echo LR\Filters::escapeHtmlText($image->getY1());
                echo '), (';
                /* line 26 */
                echo LR\Filters::escapeHtmlText($image->getX2());
                echo ', ';
                /* line 26 */
                echo LR\Filters::escapeHtmlText($image->getY2());
                echo ')
	            </li>
            </ul>

';
                /* line 30 */
                if (!$iterator->isLast()) {
                    echo '		        <hr/>
		        <hr/>
		        <hr/>';
                }
                echo "\n";
            }
            $iterator = $ʟ_it = $ʟ_it->getParent();
            echo '	';
        } finally {
            $this->global->snippetDriver->leave();
        }
    }
}
```

reports 

```
 ------ ------------------------------------------------------------------------------------ 
  Line   FrontModule/Components/Tools/TrainingImagesControl.latte                            
 ------ ------------------------------------------------------------------------------------ 
  4      Only booleans are allowed in a negated boolean, Latte\Runtime\Template|null given.  
 ------ ------------------------------------------------------------------------------------ 
```

**Comment by lulco** [comment]
@jakubvojacek I transfered the code to strict boolean comparsion 


> ```
> if (!$this->getReferringTemplate() || $this->getReferenceType() === 'extends') {
> ```
> 
> to
> 
> ```
> if ($this->getReferringTemplate() === null || $this->getReferenceType() === 'extends') {
> ```

If you'll find anything new, please create new issue. Thanks!
